src/apps/pfu.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+#include "Char.hpp"
+#include "CharVersion.hpp"
+#include <spdlog/spdlog.h>
+#include <vector>
+
+namespace pfu {
+
+class UpgradeTask {
+public:
+    explicit UpgradeTask(const CharVersion upgrade_version) : upgrade_version_(upgrade_version) {}
+    virtual ~UpgradeTask() = default;
+    [[nodiscard]] bool is_required(const Char &ch) const { return ch.version < upgrade_version_; }
+    virtual void execute(Char &ch) const = 0;
+
+private:
+    const CharVersion upgrade_version_;
+};
+
+// Reset/repair Char attributes that are modifiable via spell/item bonii.
+// This task accepts a version because over time the expectation is that
+// it will be reused for future CharVersions. It _could_ be configured to
+// use CharVersion::Max and thus run every time, but instead register_tasks()
+// will include it explicitly for versions where we need it to run.
+class ResetModifiableAttrs : public UpgradeTask {
+public:
+    ResetModifiableAttrs(CharVersion version);
+    void execute(Char &ch) const override;
+};
+
+struct CharUpgraderResult {
+    CharUpgraderResult();
+    CharUpgraderResult(Char *ch, bool upgraded);
+    ~CharUpgraderResult(); // remove the upgraded Char from the world
+    // returns true if the Char was logged in and is valid. false if it couldn't be loaded (e.g. empty file)
+    // and it's unnecessary for any upgrade tasks to have been processed because save should be called for
+    // all valid chars.
+    operator bool() const noexcept;
+    operator Char &() const noexcept;
+    Char *ch_; // set if a Char file was found, but it may still be invalid.
```

Comment:
as per the comment in the C++ file I'd set these to the defaults here and replec line 32 with `CharUpgradeRestult() = default;`. Just "feels" safer to know whatever goes on, those will be set to sensible defaults. (FUDish sounding, I know)